src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs

```diff
@@ -1,65 +1,118 @@
 			Uri authEndpoint
 		) {
 			m_httpClient = httpClient;
-			m_jwksEndpoint = BuildJwksEndpoint( authEndpoint );
+			m_authEndpoint = authEndpoint;
 		}
 		
 		async Task<JsonWebKeySet> IJwksProvider.RequestJwksAsync() {
+			var url = GetJwksEndpoint( m_authEndpoint );
+
 			try {
-				using( HttpResponseMessage response = await m_httpClient.GetAsync( m_jwksEndpoint ).SafeAsync() ) {
+				using( HttpResponseMessage response = await m_httpClient.GetAsync( url ).SafeAsync() ) {
 					response.EnsureSuccessStatusCode();
 					string jsonResponse = await response.Content.ReadAsStringAsync().SafeAsync();
-					var jwks = new JsonWebKeySet( jsonResponse, m_jwksEndpoint );
+					var jwks = new JsonWebKeySet( jsonResponse, url );
 					return jwks;
 				}
 			} catch( HttpRequestException e ) {
-				throw CreateException( e );
+				throw CreateException( e, url );
 			} catch( JsonWebKeyParseException e ) {
-				throw CreateException( e );
+				throw CreateException( e, url );
 			}
 		}
 
-		string IJwksProvider.Namespace {
-			get {
-				return m_jwksEndpoint.AbsoluteUri;
+		async Task<JsonWebKey> IJwksProvider.RequestJwkAsync( Guid keyId ) {
+			var url = GetJwkEndpoint( m_authEndpoint, keyId );
+			try {
+				using( var res = await m_httpClient.GetAsync( url ) ) {
+					// This is temporary while we try to fully deprecate the
+					// JWKS route. 404 might mean the key doesn't exist (which
+					// will make the call to jwks likely return 200 but still
+					// result in no key - that's fine but slow) or the jwk
+					// route isn't supported which will make this recover (but
+					// slowely.) Where it matters (the LMS and auth) JWK is
+					// supported so this shouldn't be necessary. We don't
+					// expect many "legitimate" 404s (keys that don't exist.)
+					// so in practice this should only happen when it's
+					// actually important, if it ever happens.
+					if( res.StatusCode == HttpStatusCode.NotFound ) {
+						return await FallbackToJwks( keyId );
+					}
+
+					res.EnsureSuccessStatusCode();
+
+					string json = await res.Content
+						.ReadAsStringAsync()
+						.SafeAsync();
+
+					return JsonWebKey.FromJson( json);
+				}
+			} catch( HttpRequestException e ) {
+				throw CreateException( e, url );
+			} catch( JsonWebKeyParseException e ) {
+				throw CreateException( e, url );
 			}
 		}
 
-		private Exception CreateException( Exception e ) {
-			string message = string.Format(
-				"Error while looking up JWKS at {0}: {1}",
-				m_jwksEndpoint,
-				e.Message
-			);
+		private async Task<JsonWebKey> FallbackToJwks( Guid keyId ) {
+			var allKeys = await ( this as IJwksProvider ).RequestJwksAsync();
+
+			JsonWebKey theKey;
+
+			if( !allKeys.TryGetKey( keyId, out theKey ) ) {
+				throw new PublicKeyLookupFailureException(
+					$"couldn't find public key {keyId} after falling back to jwks route"
+				);
+			}
+
+			return theKey;
+		}
+
+		string IJwksProvider.Namespace => m_authEndpoint.AbsoluteUri;

```

Comment:
`{ get; } =` would effectively hold a copy of that string (which is maybe just a reference internally?) i.e. it's an auto-property and equivalent to `{ get; }` with assignment in the constructor. This is equivalent to what preceded it - a regular property (no extra state.)

I don't really have a preference, I just unnecessarily applied the fixit :P